### ISSUE TITLE
Pending promises

### ISSUE THREAD

**Issue by TheLudd** [cucumber/cucumber-js#449]
With promise support, is there a way to return a promise so that a step is set to be pending?

**Comment by charlierudolph** [comment]
Nothing like that is supported at the moment. Please give more information about the use case if this is desired.

**Comment by TheLudd** [comment]
The use case is that if I want to work with promises I don't want to generate new steps with a callback argument just to have `callback.pending()`. 

If this is to be implemented, is there any other solution than returning a promise that returns a specific value? It would probably work but perhaps it would feel a bit hacky...

**Comment by mattwynne** [comment]
Can you just do `throw new Error("TODO")`?

**Comment by TheLudd** [comment]
@mattwynne Not if I want it to be pending =)

The point is that with promises there is a feature missing that is present with callbacks.

**Comment by mattwynne** [comment]
Well yes indeed! Ref https://github.com/cucumber/cucumber/issues/11 it would be interesting to hear why you care about the distinction between failing and pending.

**Comment by TheLudd** [comment]
Pending gives a much better overview than the stack traces that would be produced by returning a rejected promise. It has simply become a part of my workflow. 

 1. Write feature files 
 2. Generate snippets with pending callbacks
 3. Run and get a nice overview of what is to be done

Is the pending feature a source of pain in the codebase?

**Comment by aslakhellesoy** [comment]
@TheLudd how would your workflow be different if the generated snippets simply failed with a regular error?

**Comment by TheLudd** [comment]
To me it is mostly about readability. Pending cases are much cleaner than errors with their stack traces. Stack traces may be interesting when there actually is an error but not in the pending case.

**Comment by jbpros** [comment]
``` javascript
this.Given(/.*/, function () {
  throw new Error("Pending");
});
```

will output something like:

```
Feature: URL

  Scenario:         # features/a.feature:3
    Given something # features/step_definitions/steps.js:14
      Error: Pending
          at World.<anonymous> (/Users/jbpros/tmp/features/step_definitions/steps.js:15:11)

Failing scenarios:
features/a.feature:3 # Scenario: 

1 scenario (1 failed)
1 step (1 failed)
```

Yes, with a (quite short) stack trace.

Now throw a string instead of an error instance:

``` javascript
this.Given(/.*/, function () {
  throw "Pending";
});
```

And the stack trace will be gone:

```
Feature: URL

  Scenario:         # features/a.feature:3
    Given something # features/step_definitions/steps.js:14
      Pending

Failing scenarios:
features/a.feature:3 # Scenario: 

1 scenario (1 failed)
1 step (1 failed)
```

Now back to your original question, I can't see any other way than returning a promise that resolves (or even rejects) to a predefined value:

``` javascript
this.Given(/.*/, function () {
  // ES2015 promise:
  return new Promise(function (resolve, reject) {
    reject(Cucumber.PENDING_STEP);
  });
});
```

I don't see how this is any better than `callback.pending()`. And really, there's no need for a promise to tell Cucumber the step is pending, in the first place. There's nothing else happening in the step definition anyway, nothing asynchronous to deal with.

What would be probably better is to make Cucumber understand a predefined return value instead:

``` javascript
this.Given(/.*/, function () {
  return Cucumber.PENDING_STEP;
});
```

But then I'd rather have Cucumber understand that as a thrown value:

``` javascript
this.Given(/.*/, function () {
  throw Cucumber.PENDING_STEP;
});
```

The problem is: how do we expose that value without interfering with people's World instances?

**Comment by TheLudd** [comment]
Throwing 'Pending' worked well, thanks. I would consider returning the constant an improvement. I could consider contributing to make it work if people think it is a good idea.

**Comment by jbpros** [comment]
Thanks for offering your help @TheLudd, highly appreciated!

My suggestion was sort of pseudo-code. I'm not clear how we would expose that constant to the end users, the `Cucumber` variable isn't a real thing in a step definition. We can't use `this` within the stepdef either as it is an instance of World, managed by the user, not us. Any ideas?

One thing that could work is this:

```javascript
var Cucumber = this;

this.Given(/.*/, function () {
  throw Cucumber.PENDING_STEP;
});
```

But it's quite cumbersome.

And to be picky, not 100% correct: the thing assigned to `Cucumber` in my example is simply the object _provided_ by Cucumber exposing a few methods (`Given`, `When`, `Before`, `setDefaultTimeout`, etc.) and properties (`World`) to which we could add that `PENDING_STEP` constant.

**Comment by charlierudolph** [comment]
I'd prefer to just make the user require cucumber instead of using the context of the step definition file. We can also move the step definitions to globals and thus add another global (similar to cucumber-ruby)

**Comment by jbpros** [comment]
@charlierudolph yeah, @aslakhellesoy and I talked a bit about [using globals](https://github.com/cucumber/microcuke/commit/6930d1ecc705710ddd34a954dcc7a9411c31657b#commitcomment-13938490). Still unsure though; it's super easy to leave that to the user:

```
// features/support/env.js
module.exports = function () {
  global.Given = global.When = global.Then = this.defineStep;
};
```

The "require Cucumber" approach means we can't use that constant in stepdef snippets (because we can't force everyone to require Cucumber, just for the sake of snippets calling pending). I was thinking it'd be great to simplify the snippets and not have the callback in them anymore, once we've come up with a nicer pending signal solution.

**Comment by TheLudd** [comment]
> The "require Cucumber" approach means we can't use that constant in stepdef snippets (because we can't force everyone to require Cucumber

Well snippets are just snippets and not a complete step definitions file. You still need to export a function wrapping all steps so I wouldn't consider this a problem

**Comment by jbpros** [comment]
Sure but the code suggested by a snippet should not have implicit dependencies which would be the case if we expect people to require Cucumber (and also assign it to a `Cucumber` variable, not something else) themselves.

**Comment by charlierudolph** [comment]
This is now possible. Return the string 'pending' or return a promise that resolves to the string 'pending'.

**Comment by floribon** [comment]
@charlierudolph do you know about when are you going to do a new release with this feature?

**Comment by charlierudolph** [comment]
Wanted to wait for the next version of gherkin to be released to jump to 1.0 as this version already has a few breaking changes in it and have a couple more coming.

**Comment by charlierudolph** [comment]
0.10 released with the latest.
